cmd/kubeadm/app/phases/upgrade/versiongetter_test.go

```diff
@@ -0,0 +1,238 @@
+			}
+		})
+	}
+}
+
+func TestKubeVersionGetterVersionFromCILabel(t *testing.T) {
+	tests := []struct {
+		name              string
+		ciVersionLabel    string
+		wantCIVersion     string
+		wantParsedVersion string
+		wantErr           bool
+	}{
+		{
+			name:              "version is valid",
+			ciVersionLabel:    "v1.28.1",
+			wantCIVersion:     "v1.28.1",
+			wantParsedVersion: "1.28.1",
+			wantErr:           false,
+		},
+		{
+			name:              "version is invalid",
+			ciVersionLabel:    "invalid-version",
+			wantCIVersion:     "",
+			wantParsedVersion: "",
+			wantErr:           true,
+		},
+		{
+			name:              "version is empty",
+			ciVersionLabel:    "",
+			wantCIVersion:     "",
+			wantParsedVersion: "",
+			wantErr:           true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := &KubeVersionGetter{
+				client: clientsetfake.NewSimpleClientset(),
+			}
+			cliVersion, parsedVersion, err := g.VersionFromCILabel(tt.ciVersionLabel, "test VersionFromCILabel")
+			if (err != nil) != tt.wantErr {
+				t.Errorf("error = %v, wantErr = %v", err, tt.wantErr)
+				return
+			}
+			if cliVersion != tt.wantCIVersion {
+				t.Errorf("cliVersion = %v, wantCIVersion = %v", cliVersion, tt.wantCIVersion)
+			}
+			if tt.wantParsedVersion == "" {
+				if parsedVersion != nil {
+					t.Errorf("parsedVersion = %v, wantParsedVersion = %v", parsedVersion, tt.wantParsedVersion)
+				}
+			} else if parsedVersion.String() != tt.wantParsedVersion {
+				t.Errorf("parsedVersion = %v, wantParsedVersion = %v", parsedVersion, tt.wantParsedVersion)
+			}
+		})
+	}
+}
+
+func TestKubeVersionGetterKubeletVersions(t *testing.T) {
+	tests := []struct {
+		name    string
+		nodes   *v1.NodeList
+		want    map[string]uint16
+		wantErr bool
+	}{
+		{
+			name: "kubelet version info exists",
+			nodes: &v1.NodeList{
+				Items: []v1.Node{
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node1"},
+						Status: v1.NodeStatus{
+							NodeInfo: v1.NodeSystemInfo{
+								KubeletVersion: "v1.28.0",
+							},
+						},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node2"},
+						Status: v1.NodeStatus{
+							NodeInfo: v1.NodeSystemInfo{
+								KubeletVersion: "v1.28.1",
+							},
+						},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node3"},
+						Status: v1.NodeStatus{
+							NodeInfo: v1.NodeSystemInfo{
+								KubeletVersion: "v1.28.1",
+							},
+						},
+					},
+				},
+			},
+			want: map[string]uint16{
+				"v1.28.0": 1,
+				"v1.28.1": 2,
+			},
+			wantErr: false,
+		},
+		{
+			name: "kubelet version info is empty",
+			nodes: &v1.NodeList{
+				Items: []v1.Node{
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node2"},
+						Status:     v1.NodeStatus{},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "node3"},
+						Status:     v1.NodeStatus{},
+					},
+				},
+			},
+			want: map[string]uint16{
+				"": 2,
+			},
```

Comment:
this should not happen in reality, but the test coverage here seems okay. @neolit123 do you know is there any scenario can hit this case?

Reply:
our logic does not return errors if the key is "" - (kubeletVersion == '"")
https://github.com/my-git9/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/versiongetter.go#L123-L135

but in practice this is unlikely, because if the user edited the kubeletVersion field the node controller will update it back to the actual value. they might try building a kubelet without version info and using it for a node, but other parts of kubeadm will fail too.

@my-git9 
we can remove the `name: "kubelet version info is empty"` test case